
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Welcome to Genomeplot! &#8212; Genomeplot 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="GenomePlot" href="gp.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="welcome-to-genomeplot">
<h1>Welcome to Genomeplot!<a class="headerlink" href="#welcome-to-genomeplot" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<p><code class="docutils literal notranslate"><span class="pre">genomeplot</span></code> is a simple convenience wrapper to some <a class="reference external" href="https://bokeh.pydata.org/en/latest/">bokeh</a> functions, streamlining the creation of interactive plots across the genome.</p>
<p>It is simple to create a plot for any genome, all you need is a genome reference file (<a class="reference external" href="http://genetics.bwh.harvard.edu/pph/FASTA.html">FASTA</a> format).</p>
<p>Alternatively, this package contains a few “prerolled” genome plots, that create the base figure using the <a class="reference internal" href="gp.html#genomeplot.GenomePlot" title="genomeplot.GenomePlot"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomeplot.GenomePlot</span></code></a> class.</p>
<div class="section" id="using-a-prerolled-genomeplot">
<h2>Using a prerolled <cite>genomeplot</cite><a class="headerlink" href="#using-a-prerolled-genomeplot" title="Permalink to this headline">¶</a></h2>
<p>Instances of <code class="docutils literal notranslate"><span class="pre">GenomePlot</span></code> representing the genome of three organisms are currently available: <cite>Anopheles gambiae</cite>, <cite>Plasmodium falciparum</cite>, and <cite>Homo Sapiens</cite>.
If you would like to contribute a prerolled <code class="docutils literal notranslate"><span class="pre">GenomePlot</span></code>, please see <a class="reference internal" href="contributing.html#contributing"><span class="std std-ref">Contributing</span></a>.
The <a class="reference internal" href="gp.html#genomeplot.GenomePlot" title="genomeplot.GenomePlot"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomeplot.GenomePlot</span></code></a> instance is created using the <code class="docutils literal notranslate"><span class="pre">.load()</span></code> method. Parameters may still be edited by directly setting the values in the class.
A convenience dummy function <a class="reference internal" href="util/util.html#genomeplot.prerolled.util.noiseplot" title="genomeplot.prerolled.util.noiseplot"><code class="xref py py-func docutils literal notranslate"><span class="pre">genomeplot.prerolled.util.noiseplot()</span></code></a> is available to demonstrate <code class="docutils literal notranslate"><span class="pre">GenomePlot</span></code> instances.</p>
<div class="highlight-python notranslate" id="index.bokeh-plot-index-external-67f1ea60ba204ba783eac48b34a74daf.js"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">genomeplot</span>
<span class="kn">from</span> <span class="nn">bokeh.plotting</span> <span class="kn">import</span> <span class="n">output_file</span>

<span class="n">output_file</span><span class="p">(</span><span class="s2">&quot;example_ag.html&quot;</span><span class="p">)</span>

<span class="c1"># First create a GenomePlot instance</span>
<span class="n">agam</span> <span class="o">=</span> <span class="n">genomeplot</span><span class="o">.</span><span class="n">anophelesgambiae</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>

<span class="c1"># Pass in a function to the apply method to make a plot</span>
<span class="n">agam</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">genomeplot</span><span class="o">.</span><span class="n">util</span><span class="o">.</span><span class="n">noiseplot</span><span class="p">)</span>
</pre></div>
</div>

<script
    src="scripts/bokeh-plot-index-external-67f1ea60ba204ba783eac48b34a74daf.js"
    id="d950835d-72bd-403f-83db-d0c6c89b2dd4"
    data-bokeh-model-id="7428733b-943a-4b2c-b33a-f0633b67bbc8"
    data-bokeh-doc-id="f33ae3b3-5522-4b70-8686-e0408a33be85"
></script><div class="highlight-python notranslate" id="index.bokeh-plot-index-external-3eb9c9e73ce445f891f1d18427c4af9f.js"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">genomeplot</span>
<span class="kn">from</span> <span class="nn">bokeh.plotting</span> <span class="kn">import</span> <span class="n">output_file</span>

<span class="n">output_file</span><span class="p">(</span><span class="s2">&quot;example_hs.html&quot;</span><span class="p">)</span>

<span class="c1"># First create a GenomePlot instance</span>
<span class="n">pfal</span> <span class="o">=</span> <span class="n">genomeplot</span><span class="o">.</span><span class="n">plasmodiumfalciparum</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>

<span class="c1"># Pass in a function to the apply method to make a plot</span>
<span class="n">pfal</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">genomeplot</span><span class="o">.</span><span class="n">util</span><span class="o">.</span><span class="n">noiseplot</span><span class="p">,</span> <span class="n">winsize</span><span class="o">=</span><span class="mi">10000</span><span class="p">)</span>
</pre></div>
</div>

<script
    src="scripts/bokeh-plot-index-external-3eb9c9e73ce445f891f1d18427c4af9f.js"
    id="4367614d-2f26-46ca-b16a-4af9cd90e94d"
    data-bokeh-model-id="ba629412-b8d3-45ad-8f15-54800fa23509"
    data-bokeh-doc-id="4b5bf312-4a55-4a05-9568-3e284ad78fc9"
></script><p>In this <cite>Homo Sapiens</cite> example, we use another dummy plot function, <code class="xref py py-func docutils literal notranslate"><span class="pre">genomeplot.prerolled.util.sineplot()</span></code>, it simply plots a sine curve over each contig.</p>
<div class="highlight-python notranslate" id="index.bokeh-plot-index-external-1adf52a6cd7d4c77abf4bf38fa4d9890.js"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">genomeplot</span>
<span class="kn">from</span> <span class="nn">bokeh.plotting</span> <span class="kn">import</span> <span class="n">output_file</span>

<span class="n">output_file</span><span class="p">(</span><span class="s2">&quot;example_hs.html&quot;</span><span class="p">)</span>

<span class="c1"># First create a GenomePlot instance</span>
<span class="n">hsap</span> <span class="o">=</span> <span class="n">genomeplot</span><span class="o">.</span><span class="n">homosapiens</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>

<span class="c1"># Pass in a function to the apply method to make a plot</span>
<span class="n">hsap</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">genomeplot</span><span class="o">.</span><span class="n">util</span><span class="o">.</span><span class="n">sineplot</span><span class="p">,</span> <span class="n">mb_per_wave</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
</pre></div>
</div>

<script
    src="scripts/bokeh-plot-index-external-1adf52a6cd7d4c77abf4bf38fa4d9890.js"
    id="b4a033c1-c827-4963-863e-f8eef3cd6ecb"
    data-bokeh-model-id="bded1230-16e0-4139-9c33-8f7bd8913257"
    data-bokeh-doc-id="5ed2be9f-ba5a-45cf-ae61-3514475839a4"
></script></div>
<div class="section" id="creating-a-custom-genomeplot">
<h2>Creating a custom <cite>genomeplot</cite><a class="headerlink" href="#creating-a-custom-genomeplot" title="Permalink to this headline">¶</a></h2>
<p>First, create a new instance of the <a class="reference internal" href="gp.html#genomeplot.GenomePlot" title="genomeplot.GenomePlot"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomeplot.GenomePlot</span></code></a> class. The only required argument is a filepath describing the contigs present in the reference genome.
This is typically a <a class="reference external" href="http://genetics.bwh.harvard.edu/pph/FASTA.html">FASTA</a>, or alternatively a <code class="docutils literal notranslate"><span class="pre">csv</span></code> file, meeting the requirements described here.</p>
<p>Below is a simple example for a generic <code class="docutils literal notranslate"><span class="pre">GenomePlot</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">genomeplot</span>

<span class="n">fasta_path</span> <span class="o">=</span> <span class="s2">&quot;/path/to/fastafile.fa.gz&quot;</span>
<span class="n">contigs</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;3&quot;</span><span class="p">,</span> <span class="s2">&quot;4&quot;</span><span class="p">,</span> <span class="s2">&quot;5&quot;</span><span class="p">,</span> <span class="s2">&quot;6&quot;</span><span class="p">,</span> <span class="s2">&quot;7&quot;</span><span class="p">,</span> <span class="s2">&quot;8&quot;</span><span class="p">,</span> <span class="s2">&quot;X&quot;</span><span class="p">]</span>
<span class="n">gf</span> <span class="o">=</span> <span class="n">GenomePlot</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">,</span> <span class="n">contigs</span><span class="o">=</span><span class="n">contigs</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="s2">&quot;oo|ooo|oooo&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Once instantiated, to make a plot, a plot function is passed to the <code class="docutils literal notranslate"><span class="pre">.apply()</span></code> method of <a class="reference internal" href="gp.html#genomeplot.GenomePlot" title="genomeplot.GenomePlot"><code class="xref py py-class docutils literal notranslate"><span class="pre">genomeplot.GenomePlot</span></code></a>.
This function is then applied in turn to each contig, each making a plot which is placed in the grid.</p>
<blockquote>
<div>gf.apply(genomeplot.util.noiseplot)</div></blockquote>
<p>This function simply places a circle with a random <code class="docutils literal notranslate"><span class="pre">y</span></code> value at intervals over the genome.
Initially it is recommended that noiseplot is used as a template.
More detailed instructions to creating plotting functions suitable for use with <code class="docutils literal notranslate"><span class="pre">GenomePlot.apply()</span></code> will be available soon.
More complex examples of use will also be made available in the <a class="reference internal" href="gallery.html#gallery"><span class="std std-ref">Gallery</span></a>.</p>
</div>
<div class="section" id="source">
<h2>Source<a class="headerlink" href="#source" title="Permalink to this headline">¶</a></h2>
<p>Source code is hosted on GitHub at <a class="reference external" href="https://github.com/hardingnj/genomeplot">https://github.com/hardingnj/genomeplot</a>.</p>
</div>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>This package isn’t yet stable/mature enough to be put in <code class="docutils literal notranslate"><span class="pre">conda</span></code>/<code class="docutils literal notranslate"><span class="pre">pypi</span></code>, so for the moment please install by either cloning the repo, or downloading the <code class="docutils literal notranslate"><span class="pre">.tar.gz</span></code> from the master branch and extracting.</p>
<p>Then run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">setup</span><span class="o">.</span><span class="n">py</span> <span class="n">install</span>
</pre></div>
</div>
</div>
<div class="section" id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt>::</dt>
<dd>bokeh &gt;= 0.12.14
python &gt;= 3.5</dd>
</dl>
</div>
<div class="section" id="contributing">
<h2>Contributing<a class="headerlink" href="#contributing" title="Permalink to this headline">¶</a></h2>
<p>Pull requests for additional organisms and new features are very welcome. See <a class="reference internal" href="contributing.html#contributing"><span class="std std-ref">Contributing</span></a> for more details.</p>
</div>
<div class="section" id="license-and-warning">
<h2>License and warning<a class="headerlink" href="#license-and-warning" title="Permalink to this headline">¶</a></h2>
<p>This is academic software, has not been extensively tested, and may contain bugs and/or omissions.
If you find errors or have problems, please raise an issue at <a class="reference external" href="https://github.com/hardingnj/genomeplot/issues">https://github.com/hardingnj/genomeplot/issues</a>.
This software (including documentation) is licensed under the GNU GENERAL PUBLIC LICENSE.</p>
</div>
<div class="section" id="acknowledgements">
<h2>Acknowledgements<a class="headerlink" href="#acknowledgements" title="Permalink to this headline">¶</a></h2>
<p>Much of this code is heavily based on <a class="reference external" href="https://matplotlib.org">matplotlib</a> code written by <a class="reference external" href="https://github.com/alimanfoo">Alistair Miles</a></p>
</div>
<div class="section" id="indices-and-tables">
<h2>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Welcome to Genomeplot!</a><ul>
<li><a class="reference internal" href="#using-a-prerolled-genomeplot">Using a prerolled <cite>genomeplot</cite></a></li>
<li><a class="reference internal" href="#creating-a-custom-genomeplot">Creating a custom <cite>genomeplot</cite></a></li>
<li><a class="reference internal" href="#source">Source</a></li>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#requirements">Requirements</a></li>
<li><a class="reference internal" href="#contributing">Contributing</a></li>
<li><a class="reference internal" href="#license-and-warning">License and warning</a></li>
<li><a class="reference internal" href="#acknowledgements">Acknowledgements</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
      <li>Next: <a href="gp.html" title="next chapter">GenomePlot</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, Nicholas Harding.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.7.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>
